File: Tarea_8_PY/Libreria.py

class Libreria:
    def __init__(self):
        self.libros = []
        self.usuarios = []

    def registrar_usuario(self, usuario):
        self.usuarios.append(usuario)

    def registrar_libro(self, libro):
        self.libros.append(libro)

    def rentar_libro(self, usuario, libro):
        if not libro.is_rentado():
            usuario.libros_rentados.append(libro)
            libro.set_rentado(True)

    def listar_usuarios(self):
        return self.usuarios

    def listar_libros(self):
        return self.libros

    def listar_usuarios_con_rentas(self):
        return [usuario for usuario in self.usuarios if usuario.libros_rentados]

    def listar_libros_no_rentados(self):
        return [libro for libro in self.libros if not libro.is_rentado()]

    def listar_libros_rentados(self):
        return [libro for libro in self.libros if libro.is_rentado()]

File: Tarea_8/src/Libreria.java
import java.util.*;
public class Libreria {
    List<Libro> libros;
    List<Usuario> usuarios;

    public Libreria() {
        libros = new ArrayList<>();
        usuarios = new ArrayList<>();
    }

    void registrarUsuario(Usuario usuario) {
        usuarios.add(usuario);
    }

    void registrarLibro(Libro libro) {
        libros.add(libro);
    }

    void rentarLibro(Usuario usuario, Libro libro) {
        if (!libro.rentado) {
            usuario.librosRentados.add(libro);
            libro.rentado = true;
        }
    }

    List<Usuario> listarUsuarios() {
        return usuarios;
    }

    List<Libro> listarLibros() {
        return libros;
    }

    List<Usuario> listarUsuariosConRentas() {
        List<Usuario> resultado = new ArrayList<>();
        for (Usuario usuario : usuarios) {
            if (!usuario.librosRentados.isEmpty()) {
                resultado.add(usuario);
            }
        }
        return resultado;
    }

    List<Libro> listarLibrosNoRentados() {
        List<Libro> resultado = new ArrayList<>();
        for (Libro libro : libros) {
            if (!libro.rentado) {
                resultado.add(libro);
            }
        }
        return resultado;
    }

    List<Libro> listarLibrosRentados() {
        List<Libro> resultado = new ArrayList<>();
        for (Libro libro : libros) {
            if (libro.rentado) {
                resultado.add(libro);
            }
        }
        return resultado;
    }
}
